### DISCUSSION TITLE
are requests handled asynchronously? [QUESTION]

### DISCUSSION THREAD

**Discussion by t0gre** [fastapi/fastapi#7793] [Questions] [answered]
### Description

Firstly, I just want to start by saying I love FastAPI, it's great, and thanks for all your hard work. I am just confused about whether I should expect the api to actually handle requests asynchronously or not?

If I launch a very basic api, with file: my-api.py -

```python
import asyncio

# api 
from fastapi import FastAPI

app = FastAPI()

@app.get("/sleep/")
async def sleep(): 
    print("I'm sleepy")
    await asyncio.sleep(1)
    print("I'm awake!")
``` 
and then run it with `hypercorn my-api:app`

If I then hit it with 10 requests asynchronously, with a test script like 

```python
import requests
from datetime import datetime
import paco


URL = "http://127.0.0.1:8000/sleep/"

CALLS = 10
            
async def request_get(x):
    return requests.get(URL)


start = datetime.now()

results = paco.run(paco.map(request_get, range(CALLS)))

end = datetime.now()

print("Start time = " + start.strftime("%H:%M:%S") + "\n" + 
      "End time = " + end.strftime("%H:%M:%S") + "\n" +
      "Elapsed = " + str(end - start) + "\n" + 
      "Seconds per call = " + str((end - start)/CALLS)
      )
``` 
I get 

```bash
Start time = 16:59:18
End time = 16:59:28
Elapsed = 0:00:10.114808
Seconds per call = 0:00:01.011481
```
So, it's handling each request sequentially... is that the intended behaviour, or have I mis-configured something? 

I can work around it, but making a post route which takes a list, and then handle the list asynchronously in my code, I'm just confused about what I should be expecting here.

### Additional context

Many thanks again

**Comment by euri10** [comment]
not sure what paco is, but something like this simple shows how to run concurrently, maybe worth investigating what paco.run does under the hood

```
import asyncio

# api
from datetime import datetime

from fastapi import FastAPI

app = FastAPI()


@app.get("/sleep/")
async def sleep():
    print("I'm sleepy")
    await asyncio.sleep(1)
    print("I'm awake!")


async def main():
    start = datetime.utcnow()
    tasks = [asyncio.create_task(sleep()) for _ in range(100)]
    result = await asyncio.gather(*tasks)
    end = datetime.utcnow()
    print(end-start)
    return result


if __name__ == '__main__':
    asyncio.run(main())
```

**Comment by t0gre** [comment]
sorry, I should have said, paco is just a library for working with asyncio. If I use straight up asyncio like 

```python
import requests
from datetime import datetime
import asyncio


URL = "http://127.0.0.1:8000/sleep/"

CALLS = 10

async def request_get():
    return requests.get(URL)

async def main():
    start = datetime.now()
    tasks = [asyncio.create_task(request_get()) for _ in range(CALLS)]
    result = await asyncio.gather(*tasks)

    end = datetime.now()

    print("Start time = " + start.strftime("%H:%M:%S") + "\n" + 
        "End time = " + end.strftime("%H:%M:%S") + "\n" +
        "Elapsed = " + str(end - start) + "\n" + 
        "Seconds per call = " + str((end - start)/CALLS)
        )

asyncio.run(main())
```

I get exactly the same result. In reality I have a more complex api with lots of asyncio going on inside the api. What I can't figure out is: is the api itself supposed to handle requests asynchronously? i.e. should my  10 requests be coming back in 1 second, instead of 10 seconds? I thought that they would, but maybe I'm missing something...

**Comment by phy25** [comment]
I am not sure if requests made by `requests` is async on the client end. Could you try `httpx`?

**Comment by t0gre** [comment, marked as answer]
@phy25 thanks, that has solved it. If I replace

```python
async def request_get():
    return requests.get(URL)
```
with 

```python
async def request_get():
    async with httpx.AsyncClient() as client:
       return await client.get(URL)
```
then it executes in one second. PHEW!

I'll close the issue now. Thanks all and thanks FastAPI!

**Comment by wozniakty** [comment]
https://github.com/encode/requests-async might be of use as well

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :cake: 

Yep, HTTPX should be the preferred one :heavy_check_mark: 

Thanks @tomgreenwood1  for reporting back and closing the issue :+1:
